pkg/liqonet/ipam.go

```diff
@@ -2,108 +2,241 @@ package liqonet
+func (liqoIPAM *IPAM) GetSubnetPerCluster(network, clusterID string) (string, error) {
+	var mappedNetwork string
+	if value, ok := liqoIPAM.SubnetPerCluster[clusterID]; ok {
+		return value, nil
+	}
+	klog.Infof("Network %s allocation request for cluster %s", network, clusterID)
+	_, err := liqoIPAM.ipam.NewPrefix(network)
+	if err != nil && !strings.Contains(err.Error(), "overlaps") {
+		/* Overlapping is not considered an error in this context. */
+		return "", fmt.Errorf("Cannot reserve network %s:%w", network, err)
 	}
-	var availableSubnet *net.IPNet
-	for _, subnet := range ip.FreeSubnets {
-		availableSubnet = subnet
-		break
+	if err == nil {
+		klog.Infof("Network %s successfully assigned for cluster %s", network, clusterID)
+		liqoIPAM.SubnetPerCluster[clusterID] = network
+		return network, nil
 	}
-	return availableSubnet, nil
+	/* Since NewPrefix failed, network belongs to a pool or it has been already reserved */
+	klog.Infof("Cannot allocate network %s, checking if it belongs to any pool...", network)
+	pool, ok, err := liqoIPAM.getPoolFromNetwork(network)
+	if err != nil {
+		return "", err
+	}
+	if ok {
+		klog.Infof("Network %s belongs to pool %s, trying to acquire it...", network, pool)
+		_, err := liqoIPAM.ipam.AcquireSpecificChildPrefix(pool, network)
+		if err != nil && !strings.Contains(err.Error(), "is not available") {
+			/* Uknown error, return */
+			return "", fmt.Errorf("cannot acquire prefix %s from prefix %s: %w", network, pool, err)
+		}
+		if err == nil {
+			klog.Infof("Network %s successfully assigned to cluster %s", network, clusterID)
+			liqoIPAM.SubnetPerCluster[clusterID] = network
+			return network, nil
+		}
+		/* Network is not available, need a mapping */
+		klog.Infof("Cannot acquire network %s from pool %s", network, pool)
+	}
+	/* Network is already reserved, need a mapping */
+	klog.Infof("Looking for a mapping for network %s...", network)
+	mappedNetwork, ok = liqoIPAM.mapNetwork(network)
+	if !ok {
+		return "", fmt.Errorf("Cannot assign any network to cluster %s", clusterID)
+	}
+	klog.Infof("Network %s successfully mapped to network %s", mappedNetwork, network)
+	klog.Infof("Network %s successfully assigned to cluster %s", mappedNetwork, clusterID)
+	liqoIPAM.SubnetPerCluster[clusterID] = mappedNetwork
+	return mappedNetwork, nil
 }
 
-//add the network to the UsedSubnets and remove the subnets in free subnets that overlap with the network
-func (ip IpManager) reserveSubnet(network *net.IPNet, clusterID string) {
-	ip.UsedSubnets[network.String()] = network
-	for _, net := range ip.FreeSubnets {
-		if bool := VerifyNoOverlap(ip.UsedSubnets, net); bool {
-			ip.ConflictingSubnets[net.String()] = net
-			delete(ip.FreeSubnets, net.String())
+func (liqoIPAM *IPAM) mapNetwork(network string) (string, bool) {
```

Comment:
Can you please add a comment to this function?

Reply:
Sure!